### PR TITLE
Added trace log when crash for MSVS2013 version + homogenize trace log

### PR DESCRIPTION
**Pull request by hwoarangmy**
Tested with MinGW, MSVS2013 and Ubuntu. The linux part is the less tested so please do not hesitate do make the game crash ^^

### CODE REVIEW

**Comment by hwoarangmy** [comment]
@oyvindln Since @Bertram25 seems not to be around, could you please have a look at this ? And test a bit with linux ?
I think I am the only one to use MSVS2013 so you will have to trust me for this part ^^

**Comment by Bertram25** [comment]
Hey there, :)

I'll be around this evening to test all that. Sorry for the little comm gap. :)
/me is re-reading a lot of java at the moment, btw. XD

**Comment by hwoarangmy** [comment]
> /me is re-reading a lot of java at the moment, btw. XD

^^ May I ask why ?

**Comment by Bertram25** [comment]
> ^^ May I ask why ?

Sure. I'm on a "campus project" atm. Seems I've been accepted in a trial period for freelance tasks, here on github. Wish me good luck. ;)

**Comment by hwoarangmy** [comment]
ok. If you are at ease with C++, java should not be a problem. Good luck ^^

**Comment by Bertram25** [comment]
> ok. If you are at ease with C++, java should not be a problem. Good luck ^^

Thanks :) Actually, I already did some java but it's not my main skill. Things are coming back together already, so hopefully, I'll be back on track really soon about that.

Back on topic: I'll still review the code and test tonight.

File: CMakeLists.txt [hunk 1]
```diff
@@ -372,15 +372,23 @@ set(OD_SOURCEFILES
...
-    ${SRC}/utils/StackTraceWinMinGW.cpp
     ${SRC}/utils/Random.cpp
     ${SRC}/utils/ResourceManager.cpp
-    ${SRC}/utils/StackTracePrint.cpp
 
     ${SRC}/ODApplication.cpp
     ${SRC}/main.cpp
 )
 
+IF (MINGW)
```
Comment (Bertram25) [inline CMakeLists.txt:382]: Good! :)

File: source/entities/Tile.h [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include <string>
 #include <vector>
 #include <iosfwd>
+#include <cstdint>
```
Comment (Bertram25) [inline source/entities/Tile.h:28]: Is this actually useful? MSVC?

Comment (hwoarangmy) [inline source/entities/Tile.h:28]: yep. It is needed for MSVS2013 everywhere uintXX_t are used ^^

File: CMakeLists.txt (hunk 1 above)
Comment (hwoarangmy) [inline CMakeLists.txt:382]: Yep. @oyvindln will be happy to not have to compile uneeded files ^^

File: source/utils/StackTraceUnix.cpp [hunk 3]
```diff
@@ -20,22 +15,78 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#if not defined WIN32 && defined (__i386__) || defined (__x86_64__)  // Only for supported platforms
-
 #include "utils/StackTracePrint.h"
 
+#include <SFML/System.hpp>
```
Comment (Bertram25) [inline source/utils/StackTraceUnix.cpp]: Is that one also useful?

File: source/utils/StackTraceWinMSVC.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "utils/StackTracePrint.h"
+
+#include <SFML/System.hpp>
```
Comment (Bertram25) [inline source/utils/StackTraceWinMSVC.cpp]: Same here

File: source/utils/StackTraceStub.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "utils/StackTracePrint.h"
+
+#include <SFML/System.hpp>
```
Comment (Bertram25) [inline source/utils/StackTraceStub.cpp]: Same here. (See below)

File: source/utils/StackTraceWinMSVC.cpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+#define PACKAGE "OpenDungeon"
```
Comment (Bertram25) [inline source/utils/StackTraceWinMSVC.cpp:22]: Is PACKAGE define needed? There is a typo on OpenDungeon**s**
Same question on other files.

**Comment by Bertram25** [comment]
Looks fine otherwise. I dumbly didn't the lock using SFML, so nevermind comments about that (that were deleted since then).
I wonder whether the PACKAGE define is needed though.

**Comment by Bertram25** [comment]
I'll merge now as it is good enough, and I don't to make you lose more time on this.
Let's check about the PACKAGE afterwards. :)

File: source/utils/StackTraceWinMSVC.cpp (hunk 6 above)
Comment (hwoarangmy) [inline source/utils/StackTraceWinMSVC.cpp:22]: It is needed for MinGW bfd.h (it requires PACKAGE to be defined or it won't compile). Here, it is just a wrong copy/paste. I will remove it (and fix the typo in MinGW file).
Nice catch

**Comment by akien-mga** [comment]
@hwoarangmy Since you did a great job here to implement those backtraces in OD, I allow myself to tell you about [Lugaru](https://gitlab.com/osslugaru/lugaru), another libre C++ game I contribute to, where we could really use such a feature - in case you're looking for another project to stretch your C++ muscles :D
A warning though, it doesn't yet have the code quality we've been used to on OD...
https://gitlab.com/osslugaru/lugaru/issues/72

**Comment by hwoarangmy** [comment]
Well, it is pretty easy to do. They just have to grab StackTraceXXXX.cpp and StackTracePrint.h from our repository and use the correct one in their makefile depending on platform (as we do in ours). Then, from the point they want to start backtracing, they just have to create a StackTracePrint instance.
There is no dependency to our code excepted LogManager and Helper (which are easy to remove)

**Comment by akien-mga** [comment]
>  There is no dependency to our code excepted LogManager and Helper (which are easy to remove)

There's also `sf::Mutex`. Can we do without it? If so I'll give it a try, thanks for the explanations :)

**Comment by hwoarangmy** [comment]
The mutex is there only to ensure there is only 1 instance of the StackTracePrint object. You can remove it if you ensure the object will not be created more than once.

**Pull request merged**